apps/src/codebridge/hooks/useSource.ts

```diff
@@ -29,12 +33,25 @@ export const useSource = (defaultSources: ProjectSources) => {
   );
   const previousLevelIdRef = useRef<number | null>(null);
   const levelId = useAppSelector(state => state.lab.levelProperties?.id);
+  const isReadOnly = useAppSelector(isReadOnlyWorkspace);
+
+  const setSourceHelper = useMemo(
+    () => (newProjectSource: ProjectSources) => {
+      if (isReadOnly) {
+        // Don't attempt to save in read-only mode.
+        dispatch(setProjectSource(newProjectSource));
+      } else {
+        dispatch(setAndSaveProjectSource(newProjectSource));
+      }
+    },
```

Comment:
again, non-blocking syntactic sugar, but I think it reads a little more clearly like this:

```
const saveFunction = isReadOnly ? setProjectSource : setAndSaveProjectSource;
dispatch(saveFunction(newProjectSource);
```

Of course, if there's a change that the arguments may drift apart, then it should definitely stay the if/else